pallets/pallet-asset-swap/src/xcm/transact.rs

```diff
@@ -0,0 +1,111 @@
+use xcm_executor::traits::{ConvertLocation, TransactAsset};
+
+use crate::{Config, LocalCurrencyBalanceOf, SwapPair, SwapPairInfoOf, LOG_TARGET};
+
+// TODO: Add unit tests
+pub struct SwapPairTransactor<AccountIdConverter, T>(PhantomData<(AccountIdConverter, T)>);
+
+impl<AccountIdConverter, T> TransactAsset for SwapPairTransactor<AccountIdConverter, T>
+where
+	AccountIdConverter: ConvertLocation<T::AccountId>,
+	T: Config,
+{
+	fn deposit_asset(what: &MultiAsset, who: &MultiLocation, _context: &XcmContext) -> XcmResult {
+		// 1. Verify the swap pair exists.
+		let swap_pair = SwapPair::<T>::get().ok_or(XcmError::FailedToTransactAsset("No swap pair found."))?;
+
+		// 2. Verify the swap pair is running.
+		ensure!(
+			swap_pair.can_swap(),
+			XcmError::FailedToTransactAsset("Swap pair not enabled.")
+		);
+
+		// 3. Verify the asset matches the other side of the swap pair.
+		let stored_asset_id_as_required_version: AssetId =
+			swap_pair.remote_asset_id.clone().try_into().map_err(|e| {
+				log::error!(
+					target: LOG_TARGET,
+					"Failed to convert stored asset ID {:?} into required version with error {:?}.",
+					swap_pair.remote_asset_id,
+					e
+				);
+				XcmError::FailedToTransactAsset("Failed to convert stored asset ID into required version.")
+			})?;
+		ensure!(
+			stored_asset_id_as_required_version == what.id,
+			XcmError::FailedToTransactAsset("Unrecognized incoming asset.")
+		);
+
+		// 4. Perform the local transfer
+		let beneficiary = AccountIdConverter::convert_location(who).ok_or(XcmError::FailedToTransactAsset(
+			"Failed to convert beneficiary to valid account.",
+		))?;
+		let Fungibility::Fungible(fungible_amount) = what.fun else {
+			return Err(XcmError::FailedToTransactAsset(
+				"Deposited token expected to be fungible.",
+			));
+		};
+		let fungible_amount_as_currency_balance: LocalCurrencyBalanceOf<T> =
+			fungible_amount.try_into().map_err(|_| {
+				XcmError::FailedToTransactAsset("Failed to convert fungible amount to balance of local currency.")
+			})?;
+		T::LocalCurrency::transfer(
+			&swap_pair.pool_account,
+			&beneficiary,
+			fungible_amount_as_currency_balance,
+			Preservation::Expendable,
+		)
+		.map_err(|e| {
+			log::error!(
+				target: LOG_TARGET,
+				"Failed to transfer assets from pool account with error {:?}",
+				e
+			);
+			XcmError::FailedToTransactAsset("Failed to transfer assets from pool account")
+		})?;
+
+		// 5. Increase the balance of the remote asset
+		let new_remote_balance =
+			swap_pair
+				.remote_asset_balance
+				.checked_add(fungible_amount)
+				.ok_or(XcmError::FailedToTransactAsset(
+					"Failed to transfer assets from pool account",
+				))?;
+		SwapPair::<T>::try_mutate(|entry| {
+			let SwapPairInfoOf::<T> {
+				remote_asset_balance, ..
+			} = entry
+				.as_mut()
+				.ok_or(XcmError::FailedToTransactAsset("SwapPair should not be None."))?;
+			*remote_asset_balance = new_remote_balance;
+			Ok::<_, XcmError>(())
+		})?;
```

Comment:
I am not sure about this one. I know we checked that the XCM message can be transmitted to AssetHub, but I assume those checks are mainly local, verifying whether a channel exists or not. I am concerned that some messages on AssetHub might not be executed or might throw an error in specific edge cases. If this happens, the invariant could get corrupted.

Reply:
This cannot happen. We are 100% sure the transfer happened remotely if we are even receiving the XCM message here, so we can safely increase the counter. The opposite is not true tho, meaning that we are optimistically reducing the balance here before the full transfer of eKILTs happens on AssetHub. I thought about this. There would need to be a "ping-back" mechanism where we tell KILT, from AssetHub, that something went ok or wrong. We could use XCM message hashes, and temporarily store a map of hash -> amount locally. Every time we get a successful message, we know the transfer happened, so we are N KILTs "poorer" on AssetHub. I will mark this as something to do for the pallet V2.

Reply:
This is called when a user is trying to "reserve-deposit" eKILTs in Spiritnet. For each eKILT that AssetHub guarantees us has been moved to our sovereign account there, we release one KILT from the pool account. So the condition you are talking about has already been verified by AH before dispatching the message that we are in the middle of processing here.

Reply:
Yes, you are right. We are here 100% save. The fees will also be paid by the KILT transferred by the XCM message, so this will also not fail. 

The optimistic reducing seems also fine, but the user needs somehow a mechanism to claim back his KILT, which are not processed on AH. I am sure this will be part of v2.

Reply:
That would be an asset trap, or we can simply add instructions to deposit whatever is left after an error to the user's balance.

Reply:
This seems to me like an attack point, or not? A user can send a bunch of extrinsic, which will fail on AH.

Therefore, the `remote_asset_balance` will be so low that no one can do a swap. 

On AH, he can claim those coins and transfer them back to KILT, with no huge financial damage

Reply:
Ah no if he transfers those back the remote_asset_balance will be increased again

Reply:
Yeah, okay, he would only damage himself